kernel/frame_allocator/src/lib.rs

```diff
@@ -861,29 +1077,34 @@ fn contains_any(
     false
 }
 
-
-/// Adds the given `frames` to the given `list` as a Chunk of reserved frames. 
+/// Adds the given `frames` to the given `regions_list` and `frames_list` as a chunk of reserved frames. 
 /// 
-/// Returns the range of **new** frames that were added to the list, 
+/// Returns the range of **new** frames that were added to the lists, 
 /// which will be a subset of the given input `frames`.
 ///
 /// Currently, this function adds no new frames at all if any frames within the given `frames` list
 /// overlap any existing regions at all. 
 /// TODO: handle partially-overlapping regions by extending existing regions on either end.
-fn add_reserved_region(
-    list: &mut StaticArrayRBTree<Chunk>,
+fn add_reserved_region_to_lists(
+    regions_list: &mut StaticArrayRBTree<PhysicalMemoryRegion>,
+    frames_list: &mut StaticArrayRBTree<FreeFrames>,
     frames: FrameRange,
 ) -> Result<FrameRange, &'static str> {
 
+    // first check the regions list for overlaps and proceed only if there are none.
+    if contains_any(regions_list, &frames){
+        return Err("Failed to add reserved region that overlapped with existing reserved regions.");
+    }
```

Comment:
(this comment is for future reference)

you're close to getting the early return pattern right -- the point is to avoid the big `else` block, which is now unnecessary. See the changes I made (which also significantly reduces the noisy diff on this PR)